src/solvers/dg_triangles/dg.jl

```diff
@@ -0,0 +1,287 @@
+# !!! warning "Experimental features"
+
+# todo: replace mul! by Octavian.matmul!
+# todo: add @threaded 
+
+# out <- A*x
+mul_by!(A) = let A = A 
+  @inline (out, x)->mul!(out, A, x) 
+end
+
+# out <- out + A * x 
+mul_by_accum!(A) = let A = A 
+  @inline (out, x)->mul!(out, A, x, one(eltype(out)), one(eltype(out))) 
+end
+
+#  out <- out + α * A * x 
+mul_by_accum!(A, α) = let A = A 
+  @inline (out, x)->mul!(out, A, x, α, one(eltype(out))) 
+end
+
+const DGWeakForm{Dims, ElemType} = DG{<:RefElemData{Dims, ElemType}, Mortar, 
+                    <:SurfaceIntegralWeakForm,
+                    <:VolumeIntegralWeakForm} where {Mortar}
+
+# this is necessary for pretty printing
+Base.real(rd::RefElemData{Dims, Elem, ApproxType, Nfaces, RealT}) where {Dims, Elem, ApproxType, Nfaces, RealT} = RealT
+
+function Trixi.ndofs(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache)
+  rd = dg.basis
+  return rd.Np * mesh.md.num_elements
+end
+Trixi.wrap_array(u_ode::StructArray, semi::Trixi.AbstractSemidiscretization) = u_ode
+Trixi.wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
+
+# interface with semidiscretization_hyperbolic
+function Trixi.digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, 
+                      mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) where {Keys,ValueTypes<:Tuple{Any,Any}}
+  return boundary_conditions
+end
+
+function allocate_coefficients(mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache)
+  md = mesh.md
+  nvars = nvariables(equations) 
+  return StructArray{SVector{nvars, real(dg)}}(ntuple(_->similar(md.x),nvars))
+end
+
+function compute_coefficients!(u::StructArray, initial_condition, t,
+                 mesh::AbstractMeshData{Dim}, equations, dg::DG{<:RefElemData{Dim}}, cache) where {Dim}
+  md = mesh.md
+  rd = dg.basis
+  @unpack u_values = cache
+
+  num_quad_nodes = length(u_values) 
+  for i in Base.OneTo(num_quad_nodes) 
+    xyz_i = SVector{Dim}(getindex.(md.xyzq, i))
```

Comment:
Is this type stable and efficient?

Reply:
Yes, unless `@code_warntype` is lying to me (a la https://github.com/JuliaLang/julia/issues/32834).

Reply:
`getindex.(md.xyzq,i)` is non-allocating since `md.xyzq::NTuple{Dim,Matrix{Tv}} where {Tv}`. It returns a tuple of coordinates, and I think converting to `SVector{Dim}` should be efficient.